File: brdgme-go/splendor_1/command.go
package splendor_1

import (
	"errors"
	"fmt"

	"github.com/brdgme/brdgme/brdgme-go/brdgme"
)

type BuyCommand struct {
	Row, Col int
}

type DiscardCommand struct {
	Tokens []int
}

type ReserveCommand struct {
	Row, Col int
}

type TakeCommand struct {
	Tokens []int
}

type VisitCommand struct {
	Noble int
}

func (g *Game) Command(
	player int,
	input string,
	players []string,
) (brdgme.CommandResponse, error) {
	parser := g.CommandParser(player)
	if parser == nil {
		return brdgme.CommandResponse{}, errors.New(
			"not expecting any commands at the moment",
		)
	}
	output, err := parser.Parse(input, players)
	if err != nil {
		return brdgme.CommandResponse{}, err
	}
	switch value := output.Value.(type) {
	case BuyCommand:
		return g.BuyCommand(player, value.Row, value.Col, output.Remaining)
	case DiscardCommand:
		return g.DiscardCommand(player, value.Tokens, output.Remaining)
	case ReserveCommand:
		return g.ReserveCommand(player, value.Row, value.Col, output.Remaining)
	case TakeCommand:
		return g.TakeCommand(player, value.Tokens, output.Remaining)
	case VisitCommand:
		return g.VisitCommand(player, value.Noble, output.Remaining)
	}
	return brdgme.CommandResponse{}, errors.New("inexhaustive command handler")
}

func (g *Game) CommandSpec(player int) *brdgme.Spec {
	parser := g.CommandParser(player)
	if parser != nil {
		spec := parser.ToSpec()
		return &spec
	}
	return nil
}

func (g *Game) BuyCommand(player, row, col int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Buy(player, row, col)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) DiscardCommand(player int, tokens []int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Discard(player, tokens)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) ReserveCommand(player, row, col int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Reserve(player, row, col)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) TakeCommand(player int, tokens []int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Take(player, tokens)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) VisitCommand(player, noble int, remaining string) (brdgme.CommandResponse, error) {
	logs, err := g.Visit(player, noble)
	return brdgme.CommandResponse{
		Logs:      logs,
		CanUndo:   false,
		Remaining: remaining,
	}, err
}

func (g *Game) CommandParser(player int) brdgme.Parser {
	parsers := brdgme.OneOf{}
	if g.CanBuy(player) {
		parsers = append(parsers, g.BuyParser(player))
	}
	if g.CanDiscard(player) {
		parsers = append(parsers, g.DiscardParser(player))
	}
	if g.CanReserve(player) {
		parsers = append(parsers, g.ReserveParser(player))
	}
	if g.CanTake(player) {
		parsers = append(parsers, g.TakeParser(player))
	}
	if g.CanVisit(player) {
		parsers = append(parsers, g.VisitParser(player))
	}
	if len(parsers) == 0 {
		return nil
	}
	return parsers
}

type ParsedLoc struct {
	Row, Col int
}

func (g *Game) LocParser(player int) brdgme.Parser {
	values := []brdgme.EnumValue{}
	for row, cards := range g.Board {
		for col := range cards {
			values = append(values, brdgme.EnumValue{
				Name: fmt.Sprintf("%c%d", 'A'+col, row+1),
				Value: ParsedLoc{
					Row: row,
					Col: col,
				},
			})
		}
	}
	for col := range g.PlayerBoards[player].Reserve {
		values = append(values, brdgme.EnumValue{
			Name: fmt.Sprintf("%c4", 'A'+col),
			Value: ParsedLoc{
				Row: 3,
				Col: col,
			},
		})
	}
	return brdgme.Enum{
		Values: values,
	}
}

func TokenParser(includeGold bool) brdgme.Parser {
	tokens := append([]int{}, Gems...)
	if includeGold {
		tokens = append(tokens, Gold)
	}
	values := make([]brdgme.EnumValue, len(tokens))
	for i, t := range tokens {
		values[i] = brdgme.EnumValue{
			Name:  ResourceStrings[t],
			Value: t,
		}
	}
	return brdgme.Enum{
		Values: values,
	}
}

func TokensParser(includeGold bool) brdgme.Parser {
	var tokensParserMin = uint(1)
	return brdgme.Map{
		Parser: brdgme.Many{
			Parser: TokenParser(includeGold),
			Delim:  brdgme.Space{},
			Min:    &tokensParserMin,
		},
		Func: func(value interface{}) interface{} {
			tokens := []int{}
			for _, t := range value.([]interface{}) {
				tokens = append(tokens, t.(int))
			}
			return tokens
		},
	}
}

func (g *Game) BuyParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "buy",
				Desc:   "buy a card",
				Parser: brdgme.Token("buy"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "card",
					Desc:   "the card to buy",
					Parser: g.LocParser(player),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			loc := value.([]interface{})[1].(ParsedLoc)
			return BuyCommand(loc)
		},
	}
}

func (g *Game) DiscardParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "discard",
				Desc:   "discard tokens back down to 10",
				Parser: brdgme.Token("discard"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "tokens",
					Desc:   "the tokens to discard",
					Parser: TokensParser(true),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return DiscardCommand{
				Tokens: value.([]interface{})[1].([]int),
			}
		},
	}
}

func (g *Game) ReserveParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "reserve",
				Desc:   "reserve a card and take a gold",
				Parser: brdgme.Token("reserve"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "card",
					Desc:   "the card to reserve",
					Parser: g.LocParser(player),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			loc := value.([]interface{})[1].(ParsedLoc)
			return ReserveCommand(loc)
		},
	}
}

func (g *Game) TakeParser(player int) brdgme.Parser {
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "take",
				Desc:   "take 3 different tokens, or 2 of the same token",
				Parser: brdgme.Token("take"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name:   "tokens",
					Desc:   "the tokens to take",
					Parser: TokensParser(false),
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return TakeCommand{
				Tokens: value.([]interface{})[1].([]int),
			}
		},
	}
}

func (g *Game) VisitParser(player int) brdgme.Parser {
	min := 1
	max := len(g.Nobles)
	return brdgme.Map{
		Parser: brdgme.Chain{
			brdgme.Doc{
				Name:   "visit",
				Desc:   "visit a noble",
				Parser: brdgme.Token("visit"),
			},
			brdgme.AfterSpace(
				brdgme.Doc{
					Name: "noble",
					Desc: "the noble to visit",
					Parser: brdgme.Int{
						Min: &min,
						Max: &max,
					},
				},
			),
		},
		Func: func(value interface{}) interface{} {
			return VisitCommand{
				Noble: value.([]interface{})[1].(int) - 1,
			}
		},
	}
}

File: rust/game/splendor-2/src/command.rs
//! Ported from `brdgme-go/splendor_1/command.go`, `take_command.go`,
//! `buy_command.go`, `reserve_command.go`, `discard_command.go`,
//! `visit_command.go`.

use brdgme_game::command::parser::*;

use crate::Game;
use crate::card::{GEMS, Resource};

/// Ported from `command.go`'s `ParsedLoc`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ParsedLoc {
    pub row: usize,
    pub col: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Command {
    Buy(ParsedLoc),
    Discard(Vec<Resource>),
    Reserve(ParsedLoc),
    Take(Vec<Resource>),
    Visit(usize),
}

/// A loc choice paired with its positional letter+row name, for use with the
/// `Enum` parser. Port of `LocParser`'s `EnumValue`s (`command.go`).
#[derive(Debug, Clone, Copy)]
struct LocChoice {
    loc: ParsedLoc,
    name: [u8; 2],
}

impl std::fmt::Display for LocChoice {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", self.name[0] as char, self.name[1] as char)
    }
}

/// A resource token choice paired with its `ResourceStrings` name, for use
/// with the `Enum` parser. Port of `TokenParser`'s `EnumValue`s
/// (`command.go`).
#[derive(Debug, Clone, Copy)]
struct TokenChoice {
    resource: Resource,
}

impl std::fmt::Display for TokenChoice {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.resource.name())
    }
}

impl Game {
    pub fn command_parser(&self, player: usize) -> Option<Box<dyn Parser<T = Command> + '_>> {
        let mut parsers: Vec<Box<dyn Parser<T = Command>>> = vec![];
        // Fixed order per `CommandParser` (`command.go`): buy, discard,
        // reserve, take, visit.
        if self.can_buy(player) {
            parsers.push(Box::new(self.buy_parser(player)));
        }
        if self.can_discard(player) {
            parsers.push(Box::new(discard_parser()));
        }
        if self.can_reserve(player) {
            parsers.push(Box::new(self.reserve_parser(player)));
        }
        if self.can_take(player) {
            parsers.push(Box::new(take_parser()));
        }
        if self.can_visit(player) {
            parsers.push(Box::new(self.visit_parser()));
        }
        if parsers.is_empty() {
            None
        } else {
            Some(Box::new(OneOf::new(parsers)))
        }
    }

    /// Port of `LocParser` (`command.go`): board locations named
    /// `{'A'+col}{row+1}`, then the player's own reserve locations named
    /// `{'A'+col}4`. Computed fresh from current board/reserve state on
    /// every call - column letters are positional, not stable ids.
    fn loc_parser(&self, player: usize) -> impl Parser<T = ParsedLoc> {
        let mut values: Vec<LocChoice> = vec![];
        for (row, cards) in self.board.iter().enumerate() {
            for col in 0..cards.len() {
                values.push(LocChoice {
                    loc: ParsedLoc { row, col },
                    name: [b'A' + col as u8, b'1' + row as u8],
                });
            }
        }
        for col in 0..self.player_boards[player].reserve.len() {
            values.push(LocChoice {
                loc: ParsedLoc { row: 3, col },
                name: [b'A' + col as u8, b'4'],
            });
        }
        Map::new(Enum::exact(values), |c: LocChoice| c.loc)
    }

    pub fn buy_parser(&self, player: usize) -> impl Parser<T = Command> {
        Map::new(
            Chain2::new(
                Doc::name_desc("buy", "buy a card", Token::new("buy")),
                AfterSpace::new(Doc::name_desc(
                    "card",
                    "the card to buy",
                    self.loc_parser(player),
                )),
            ),
            |(_, loc)| Command::Buy(loc),
        )
    }

    pub fn reserve_parser(&self, player: usize) -> impl Parser<T = Command> {
        Map::new(
            Chain2::new(
                Doc::name_desc(
                    "reserve",
                    "reserve a card and take a gold",
                    Token::new("reserve"),
                ),
                AfterSpace::new(Doc::name_desc(
                    "card",
                    "the card to reserve",
                    self.loc_parser(player),
                )),
            ),
            |(_, loc)| Command::Reserve(loc),
        )
    }

    pub fn visit_parser(&self) -> impl Parser<T = Command> {
        // Port of `VisitParser` (`command.go`): accepts any of `len(Nobles)`
        // by 1-based index, not just currently-affordable ones (quirk 1).
        Map::new(
            Chain2::new(
                Doc::name_desc("visit", "visit a noble", Token::new("visit")),
                AfterSpace::new(Doc::name_desc(
                    "noble",
                    "the noble to visit",
                    Int {
                        min: Some(1),
                        max: Some(self.nobles.len() as i32),
                    },
                )),
            ),
            |(_, noble)| Command::Visit((noble - 1) as usize),
        )
    }
}

/// Port of `TokenParser` (`command.go`).
fn token_parser(include_gold: bool) -> impl Parser<T = Resource> {
    let mut values: Vec<TokenChoice> = GEMS.iter().map(|&r| TokenChoice { resource: r }).collect();
    if include_gold {
        values.push(TokenChoice {
            resource: Resource::Gold,
        });
    }
    Map::new(Enum::exact(values), |c: TokenChoice| c.resource)
}

/// Port of `TokensParser` (`command.go`): one-or-more space-delimited tokens.
fn tokens_parser(include_gold: bool) -> impl Parser<T = Vec<Resource>> {
    Many::some_spaced(token_parser(include_gold))
}

pub fn take_parser() -> impl Parser<T = Command> {
    Map::new(
        Chain2::new(
            Doc::name_desc(
                "take",
                "take 3 different tokens, or 2 of the same token",
                Token::new("take"),
            ),
            AfterSpace::new(Doc::name_desc(
                "tokens",
                "the tokens to take",
                tokens_parser(false),
            )),
        ),
        |(_, tokens)| Command::Take(tokens),
    )
}

pub fn discard_parser() -> impl Parser<T = Command> {
    Map::new(
        Chain2::new(
            Doc::name_desc(
                "discard",
                "discard tokens back down to 10",
                Token::new("discard"),
            ),
            AfterSpace::new(Doc::name_desc(
                "tokens",
                "the tokens to discard",
                tokens_parser(true),
            )),
        ),
        |(_, tokens)| Command::Discard(tokens),
    )
}
